### DISCUSSION TITLE
Prisma, Typescript, Jest - Unit Testing

### DISCUSSION THREAD

**Discussion by maralerdene** [prisma/prisma#13646] [Q&A]
Hello Dears,

I'm trying to unit test with Jest following this link https://www.prisma.io/docs/guides/testing/unit-testing.
Here is my code: 
```ts
test("should create a user ", async () => {
  var user: any = {
    store_id: "601d6dea-9b89-4c0d-adfc-6ed52424a2a0",
    password: '1234',
    user_name: "username",
    first_name: "Rich",
    last_name: "hello@prisma.io",
    phone: "12345",
    is_phone_verified: true,
    is_email_verified: true,
  };

  prismaMock.sec_user_fs.create.mockImplementation(user)

  await expect(createUser(user)).resolves.toEqual({
    store_id: "601d6dea-9b89-4c0d-adfc-6ed52424a2a0"
  });
});

```

createUser method is save data to database. Here is code: 
```ts
export async function createUser(user: CreateUser) {
  
    user = await prisma.sec_user_fs.create({
      data: user,
    })

    return {"store_id": user.store_id}
}
```

So my questions are:

1. prismaMock.sec_user_fs.create.mockImplementation(user) - In my understanding this line mocked database table, it does not insert data to table. But, it does. It is possible that can mock without inserting to table?

2. createUser method return json object. But if i test my api end point /code is below/ how expect result? Because it does not return instead res.status(200) etc.?

```ts
export default asyncHandler(async (req: CustomRequest, res: Response) => {
    
   
      const body = {
        user_name: req.body.user_name,
        first_name: req.body.first_name,
        last_name: req.body.last_name,
        phone: req.body.phone,
        phone_country_code: req.body.phone_country_code
      }
  
      const user_id = req.params.id
  
      const user = await prisma.sec_user_fs.updateMany({
        where: {
          user_id,
          status: 'A'
        },
        data: body
      });
  
      res.status(200).send({
        status: 'Success',
        message: req.t('success'),
        errors: [],
        data: user,
      });
  
  });
```

Great thank you in advance :)

**Comment by austincrim** [comment]
Hey @maralerdene 👋,

Just to double check, have you followed the prerequisite steps as well? Specifically, the `setupFilesAfterEnv` option.

<img width="698" alt="CleanShot 2022-06-06 at 14 39 56@2x" src="https://user-images.githubusercontent.com/32459922/172235250-893dc652-277c-4372-bbc3-188a8614fab2.png">

**Comment by maralerdene** [comment]
Hello @austincrim 

Yes, I do. Here is my jest.config.js:

```
module.exports = {
    clearMocks: true,
    preset: 'ts-jest',
    testEnvironment: 'node',
    setupFilesAfterEnv: ['singleton.ts'],
}
```

My package.json:

```
{
  "name": "front-service",
  "version": "1.0.0",
  "main": "app.js",
  "license": "MIT",
  "scripts": {
    "start": "nodemon  -r esm target/app.js",
    "dev": "nodemon  -r esm src/app.ts",
    "build": "tsc -p .",
    "test": "jest -i"
  },
  "dependencies": {
    "@prisma/client": "^3.14.0",
    "@types/bcryptjs": "^2.4.2",
    "@types/cors": "^2.8.12",
    "@types/express-useragent": "^1.0.2",
    "@types/http-errors": "^1.8.2",
    "@types/i18next-fs-backend": "^1.1.2",
    "@types/jest": "^28.1.0",
    "@types/lodash": "^4.14.182",
    "@types/mailgun-js": "^0.22.12",
    "@types/morgan": "^1.9.3",
    "axios": "^0.27.2",
    "bcryptjs": "^2.4.3",
    "cors": "^2.8.5",
    "cote": "^1.0.3",
    "dotenv": "^16.0.0",
    "email-validator": "^2.0.4",
    "esm": "^3.2.25",
    "express": "^4.17.2",
    "express-async-handler": "^1.2.0",
    "express-useragent": "^1.0.15",
    "express-validator": "^6.14.0",
    "firebase-admin": "^10.0.2",
    "i18next": "^21.6.16",
    "i18next-fs-backend": "^1.1.4",
    "i18next-http-middleware": "^3.2.0",
    "jest": "^28.1.0",
    "jest-mock-extended": "2.0.4",
    "jsonwebtoken": "^8.5.1",
    "lodash": "^4.17.21",
    "mailgun-js": "^0.22.0",
    "morgan": "^1.10.0",
    "moxios": "^0.4.0",
    "passport": "^0.5.2",
    "passport-jwt": "^4.0.0",
    "prisma": "^3.14.0",
    "rotating-file-stream": "^3.0.3",
    "sinon": "^14.0.0",
    "supertest": "^6.2.3",
    "ts-jest": "^28.0.4",
    "ts-node": "^10.7.0",
    "typescript": "^4.6.4"
  },
  "devDependencies": {
    "@types/express": "^4.17.13",
    "@types/jsonwebtoken": "^8.5.8",
    "@types/node": "^17.0.25",
    "nodemon": "^2.0.15",
    "typescript": "^4.6.4"
  }
}

```

**Reply by austincrim** [reply]
Can you share the contents of `singleton.ts`?

**Reply by maralerdene** [reply]
Here is singleton.ts 
```ts
import { PrismaClient } from '@prisma/client'
import { mockDeep, mockReset, DeepMockProxy } from 'jest-mock-extended'
import prisma from './client'

jest.mock('./client', () => ({
  __esModule: true,
  default: mockDeep<PrismaClient>(),
}))

beforeEach(() => {
  console.log('resetted')
  mockReset(prismaMock)
})

export const prismaMock = prisma as unknown as DeepMockProxy<PrismaClient>
```


Also here is client.ts 

```ts
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()
export default prisma
```

**Reply by austincrim** [reply]
Are you certain that `singleton.ts` is getting picked up by Jest? 

Is the path you have specified in `setupFilesAfterEnv` correct?

Does the `console.log('resetted')` show up when running your tests?

**Reply by maralerdene** [reply]
Yes it works. Here is my test result. It looks passed, but inserted data to table. My question is how can i mock my table without inserting data to table?

![image](https://user-images.githubusercontent.com/8305143/172742531-391aef34-ccd9-4c4b-9620-2508a541ef68.png)

**Reply by austincrim** [reply]
It looks like your setup should be working. The easiest way for me to continue investigating would be if you provided me a reproduction repo of your exact setup.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
